### PR TITLE
feat: rust plugin rename rust keywords with trail underscore

### PR DESCRIPTION
**Pull request by driftluo**
This PR Fix two bugs on molecule：

1. rust keywords on table/struct can generate code, but cannot compile with rustc, for example:

```mol
array struct [byte; 2];

struct StructP {
    type: struct
}
```

It is OK to name a type as a rust keyword; the rust plugin will automatically camelCase the name. However, when the field of a table/struct is named as a rust keyword, the current version does not automatically handle this problem, causing the generated code to fail to compile. This PR will automatically check the field name and add an underscore to the end of the keyword to avoid keyword conflicts.

Just like the `type` above, in the generated code, it will become `type_`(Or prefix it with `r#`? I'm not sure)

**I think the solution to keyword conflicts must be solved in each language's plugins.**

2. Since type names support uppercase and lowercase letters, they will be uniformly camel-cased. When two types have inconsistent names but consistent camel case, uncompilable code will be generated, as in the following schema:

```
array struct [byte; 2];

struct StructP {
    a: struct
}

union structP {
    struct,
}
```
This PR will detect this situation and report an error: "the name `structP` is used more than once, It seems that only the capitalization is inconsistent"

other change: upgrade serde_yaml/clap/pest deps

### CODE REVIEW

File: tools/codegen/src/generator/languages/rust/utilities.rs [hunk 1]
```diff
@@ -1,15 +1,97 @@
...
+fn avoid_keywords(name: &str) -> String {
+    let keywords = rust_keyword();
+    let snake_name = name.to_snake();
+    if keywords.contains(&snake_name) {
+        snake_name + "_"
+    } else {
+        snake_name
+    }
+}
+
```
Comment (eval-exec) [inline tools/codegen/src/generator/languages/rust/utilities.rs:85]: If the user inputs `Struct` as a field name, the function converts it to lowercase and appends an underscore, changing it to `struct_`.

Perhaps it doesn't need to be converted to lowercase; appending `_` is sufficient, so it should just become `Struct_`?

**Review by eval-exec — DISMISSED** [review]

**Review by eval-exec — DISMISSED** [review]

**Comment by zhangsoledad** [comment]
MSRV was bumped because of the introduction of https://doc.rust-lang.org/std/sync/struct.OnceLock.html, this should have been stated in the README

File: tools/codegen/src/generator/languages/rust/utilities.rs [hunk 2]
```diff
@@ -1,15 +1,97 @@
 use case::CaseExt;
 use proc_macro2 as m4;
 
+use std::{collections::HashSet, sync::OnceLock};
+
+static RUST_KEYWORDS: OnceLock<HashSet<String>> = OnceLock::new();
```
Comment (quake) [inline tools/codegen/src/generator/languages/rust/utilities.rs:6]:
```suggestion
static RUST_KEYWORDS: OnceLock<HashSet<&'static str>> = OnceLock::new();
```

File: tools/codegen/src/generator/languages/rust/utilities.rs [hunk 3]
```diff
@@ -1,15 +1,97 @@
...
+            "yield",
+            // 2018+
+            "try",
+            // weak
+            "macro_rules",
+            "union",
+            "'static",
+        ])
+        .map(ToString::to_string)
+        .collect()
```
Comment (quake) [inline tools/codegen/src/generator/languages/rust/utilities.rs:72]:
```suggestion
        .collect()
```

File: tools/codegen/src/generator/languages/rust/utilities.rs (hunk 1 above)
Comment (doitian) [reply tools/codegen/src/generator/languages/rust/utilities.rs:85]: Then the function name `avoid_keywords` does not make sense.

Comment (eval-exec) [reply tools/codegen/src/generator/languages/rust/utilities.rs:85]: I asked GPT-4, and it suggested renaming it to `sanitize_identifier`.

**Review by eval-exec — APPROVED** [review]

**Review by doitian — APPROVED** [review]

**Review by doitian — APPROVED** [review]

**Review by eval-exec — APPROVED** [review]

**Pull request merged**